feat: add docs for passing test content to the Test Workflows (#13)
kubeshop/testkube-docs: docs/articles/test-workflows-content.md

import Tabs from "@theme/Tabs";
import TabItem from "@theme/TabItem";

# Test Workflows - Content

Often you need to provide some input data for your tests - whether it's Tests source code,
or some fixtures.

:::note

All the examples here that are using `content` are using it directly under the `spec`.

> Please note, that you can also use it under specific step - this way, such files won't be available in different steps.

Reviewer: ```suggestion
Please note, that you can also use it under a specific step - this way, such files won't be available in different steps.
```

:::

## Shared Directories

### /data

> By default, `/data` directory has an empty volume mounted, that can be shared across the steps.

Reviewer: ```suggestion
By default, the `/data` directory has an empty volume mounted that can be shared across the steps.
```

:::info

Please note, that this directory is shared across steps of a single execution. To share data across multiple executions, see [**Shared Directory Between Executions**](#shared-directory-between-executions).

:::

### Custom Volumes

You may create different directories that will share data across steps using Kubernetes. To do so, create and mount [**Empty Dir**](https://kubernetes.io/docs/concepts/storage/volumes/#emptydir) volume.

> This may be useful for i.e. sharing a dependencies across steps.

Reviewer: ```suggestion
This may be useful for sharing a dependencies across steps, for example.
```

```yaml
spec:
  pod:
    volumes:
    - name: golang-deps
      emptyDir: {}
  container:
    env:
    - name: GOCACHE
      value: /go-cache
    volumeMounts:
    - name: golang-deps
      mountPath: /go-cache
```

### Shared Directory Between Executions

:::info

> We are planning to prepare built-in caching mechanism, that will allow you to share data more easily.

Reviewer: ```suggestion
We are planning to prepare a built-in caching mechanism that will allow you to share data more easily.
```

:::

> If you want to cache some data between executions, i.e. to speed up execution, you may consider mounting [**Host's file system**](https://kubernetes.io/docs/concepts/storage/volumes/#hostpath) with `hostPath`.

Reviewer: ```suggestion
If you want to cache some data between executions, i.e., to speed up execution, you may consider mounting [**Host's file system**](https://kubernetes.io/docs/concepts/storage/volumes/#hostpath) with `hostPath`.
```

```yaml
spec:
  pod:
    volumes:
    - name: golang-deps
      hostPath:
        path: '/tmp/go-cache-{{ workflow.name }}'
  container:
    env:
    - name: GOCACHE
      value: /go-cache
    volumeMounts:
    - name: golang-deps
      mountPath: /go-cache
```

This way, you will have a cache for each execution of the Test Workflow on the same node.

:::tip

> If you want to cache data across multiple nodes, you may consider to:

Reviewer: ```suggestion
If you want to cache data across multiple nodes, you may consider:
```

> * use Object Storage API like [**Minio**](https://min.io/), [**AWS S3**](https://aws.amazon.com/s3/) or [**GCP Cloud Storage**](https://cloud.google.com/storage),

Reviewer: ```suggestion
* using an Object Storage API like [**Minio**](https://min.io/), [**AWS S3**](https://aws.amazon.com/s3/) or [**GCP Cloud Storage**](https://cloud.google.com/storage),
```

  and save/read files using it, or

> * create [**Persistent Volume**](https://kubernetes.io/docs/concepts/storage/persistent-volumes/) and attach it with `pod.volumes` and `container.volumeMounts`.

Reviewer: ```suggestion
* creating a [**Persistent Volume**](https://kubernetes.io/docs/concepts/storage/persistent-volumes/) and attaching it with `pod.volumes` and `container.volumeMounts`.
```

          key: key-name-in-secret
    - path: /some/path/nginx.conf
      contentFrom:
        configMapKeyRef: # Mount from ConfigMap
          name: secret-name
          key: key-name-in-secret
```

### Mounting Secret or ConfigMap as a Directory

As an alternative, you may use native Kubernetes volumes for mounting [**Secrets**](https://kubernetes.io/docs/concepts/storage/volumes/#secret) or [**ConfigMaps**](https://kubernetes.io/docs/concepts/storage/volumes/#configmap).

```yaml
spec:
  pod:
    volumes:
    # Create volume from the Secret
    - name: example-secret-volume
      secret:
        secretName: secret-name
    # Create volume from the ConfigMap
    - name: example-configmap-volume
      configMap:
        name: configmap-name
  container:
    volumeMounts:
    # Mount files from the Secret
    - name: example-secret-volume
      mountPath: /some/secret
    # Mount files from the ConfigMap
    - name: example-configmap-volume
      mountPath: /some/config
```

## Git Repository

Testkube allows you to easily fetch the Git repository using `content.git`.

### Custom Mount Path

By default, the Git repository contents are mounted at `/data/repo` directory.

> If you want to mount it in different directory, you can use `mountPath` property:

Reviewer: ```suggestion
If you want to mount it in a different directory, you can use `mountPath` property:
```

The `username` and `token` properties take plain-text arguments. If you want to take username and password/token from the Secret or ConfigMap,
you can use `usernameFrom` and/or `tokenFrom` properties. They have the same interface as `valueFrom` in [**native EnvVar**](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.23/#envvarsource-v1-core).

```yaml
spec:
  content:
    git:
      uri: https://github.com/kubeshop/testkube.git
      username: my-username
      # usernameFrom:
      #   secretKeyRef:
      #     name: git-credentials
      #     key: username
      token: my-token
      # tokenFrom:
      #   secretKeyRef:
      #     name: git-credentials
      #     key: token
```

### Using Authorization Header

The example above is using Basic Authorization. Some Git providers may require Bearer Authentication instead.
To provide the bearer token, you need to pass `authType: header` property.

```yaml
spec:
  content:
    git:
      uri: https://github.com/kubeshop/testkube.git
      authType: header
      token: my-token
      # tokenFrom:
      #   secretKeyRef:
      #     name: git-credentials
      #     key: token
```

### Using SSH Key

> You can use private SSH key to access your repository, using either plain-text `sshKey` property,

Reviewer: ```suggestion
You can use a private SSH key to access your repository, using either plain-text `sshKey` property,
```

or `sshKeyFrom` that has the same interface as `valueFrom` in [**native EnvVar**](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.23/#envvarsource-v1-core).

```yaml
spec:
  content:
    git:
      uri: ssh://git@github.com/kubeshop/testkube.git
      sshKeyFrom:
        secretKeyRef:
          name: git-credentials
          key: ssh-key
      # sshKey: |
      #   -----BEGIN OPENSSH PRIVATE KEY-----
      #   b3BlbnNzaC1rZXktdjEAAAAABG5vbmUAAAAEbm9uZQAAAAAAAAABAAAAMwAAAAtzc2gtZW
      #   XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
      #   XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
      #   XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
      #   Py55nRNObUBBaG/XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX=
      #   -----END OPENSSH PRIVATE KEY-----
```

### Pulling Selected Files

> You may pass file patterns ([**cone pattern set**](https://git-scm.com/docs/git-sparse-checkout#_internalscone_pattern_set)) in `content.git.paths` that should be fetched from the Git repository.

Reviewer: ```suggestion
You may pass file patterns ([**cone pattern set**](https://git-scm.com/docs/git-sparse-checkout#_internalscone_pattern_set)) in `content.git.paths` which should be fetched from the Git repository.
```

```yaml
spec:
  config:
    branch:
      type: string
      default: main
  content:
    git:
      uri: https://github.com/kubeshop/testkube.git
      revision: '{{ config.branch }}'
```

## Fetching Tarballs

Testkube have a simple built-in way to fetch tarball archives for your tests. You may use `content.tarball` to specify the files to download and unpack:

```yaml
spec:
  content:
    tarball:
    - url: https://github.com/kubeshop/testkube/releases/download/v1.17.53/testkube_1.17.53_Linux_arm64.tar.gz
      path: /data/bin
  steps:
  - shell: tree /data/bin
```

## Fetching OCI Artifacts

Sometimes, especially in air-gapped environment, you may want to provide the test data via OCI Registries.

You may use [**ORAS**](https://oras.land/) to fetch artifacts from the OCI registry.
To do so, run a step with Docker image that has ORAS installed, and pull the artifact into `/data` (or other known volume):

```yaml
spec:
  steps:
  - run:
      image: bitnami/oras:latest
      workingDir: /data
      shell: |
        oras pull 12345678901234.dkr.ecr.us-west-2.amazonaws.com/some/artifact:latest
  - shell: tree /data
```

### OCI Images

> Sometimes the registry won't support the OCI Artifacts, and instead they are available as regular images.

Reviewer: ```suggestion
Sometimes the registry won't support the OCI Artifacts, and, instead, they are available as regular images.
```
